---
title: About this workshop
---

This workshop focuses on showing a simplified workflow
to mine phages in a metagenome.

We will use:

1. A viral miner tool to predict viral sequences in a metagenome.
2. We will use a program to check the quality of the predictions.
3. We will select some of the predictions as candidate "Viral OTUs" (vOTUs).

Finally, we will:
4. Back-map the reads to the vOTUs to estimate their abundance in the original samples.
5. Use Anvi'o to generate a visualization focused on the vOTUs, building a CONTIGS.db database only based on the candidate vOTUs.

## EBAME settings

We will use VMs provided by Biosphere, and with a setup script datasets and databases will be automatically available.
See [EBAME Setup]({{ site.baseurl }}{% link 2000-02-01-ebame.md  | relative_url %})