---
title: About this workshop
---

This workshop focuses on showing a simplified workflow
to mine phages in a metagenome.

Starting from reads, we need to do some quality filtering and assembly to get contigs.

In a simplified workflow:

![workflow]({{ site.baseurl }}{% link img/our_wf.png %})

We will start with the co-assembly of three samples, so we will use:

1. A viral miner tool to predict viral sequences in a metagenome (geNomad).
2. A program to check the quality of the predictions (checkV).
3. finally, select some of the predictions as candidate "Viral OTUs" (vOTUs).

To keep our Anvi'o love high, we can also:

4. Back-map the reads to the vOTUs to estimate their abundance in the original samples.
5. Use Anvi'o to generate a visualization focused on the vOTUs, building a CONTIGS.db database only based on the candidate vOTUs.

## EBAME settings

We will use VMs provided by Biosphere, and with a setup script datasets and databases will be automatically available.
See [EBAME Setup]({{ site.baseurl }}{% link modules/Intro/_posts/2000-02-01-ebame.md  %})

