---
title: SeqFu
published: true
---

<img alt="seqfu" src="https://github.com/telatin/seqfu2/raw/main/docs/img/seqfu-512.png" width="256" height="256">


SeqFu is a tool to help you manage your sequence files. It is a command-line tool that can be used to filter, transform, and manipulate sequence files in various formats. SeqFu is designed to be fast and efficient, and it is written in Rust.

## Some functions

SeqFu has a set of subtools like:

* [`seqfu stats`](https://telatin.github.io/seqfu2/tools/stats.html) to get statistics of FASTA and FASTQ files, including N50 (supports gzipped files)
* [`seqfu head`](https://telatin.github.io/seqfu2/tools/head.html) and [`seqfu tail`](https://telatin.github.io/seqfu2/tools/tail.html) will print the first or last sequences. With options to print one every *N* and more.
* [`seqfu cat`](https://telatin.github.io/seqfu2/tools/cat.html), to reshape FASTA/FASTQ files, adding prefixes or suffixes to the names, removing comments, etc.
* [`seqfu interleave`](https://telatin.github.io/seqfu2/tools/interleave.html) and [`seqfu deinterleave`](https://telatin.github.io/seqfu2/tools/deinterleave.html)
  
and [many other tools](https://telatin.github.io/seqfu2/)!

## Installation

You can install if from conda

```bash
conda install -y -c conda-forge -c bioconda seqfu
```

Alternatively, you can download pre-compiled binaries and put  them on your PATH. 
Note that in the EBAME VM you have SeqFu in `~/bin/`.

## Anvi'o special

```bash
seqfu cat --anvio ASSEMBLY > RENAMED_ASSEMBLY.fa
```

will substitute `anvi-script-reformat-fasta`. It's faster, and being SeqFu easy to install, you can embed this step
in your pipeline to get Anvi'o ready contigs.

## Visual commands

* **[seqfu view](https://telatin.github.io/seqfu2/tools/view.html)**

You can have a quick look at FASTQ files (to check for quality and primers/adapters):

```bash
seqfu cat FASTQ_FILE | less -SR
```

![view](https://telatin.github.io/seqfu2/img/screenshot-view-example.svg)


* **[fu-msa](https://telatin.github.io/seqfu2/utilities/fu-msa.htmll)**

To check multiple alignment files from the command line:

```bash
fu-msa MSA.fa
```

![view](https://telatin.github.io/seqfu2/img/msa.png)